ntp-proto/src/clock.rs

```diff
@@ -8,5 +10,397 @@ pub trait NtpClock {
+    fn set_freq(&self, freq: f64) -> Result<(), Self::Error>;
+    fn step_clock(&self, offset: NtpDuration) -> Result<(), Self::Error>;
+    fn update_clock(
+        &self,
+        offset: NtpDuration,
+        est_error: NtpDuration,
+        max_error: NtpDuration,
+        poll_interval: PollInterval,
+        leap_status: NtpLeapIndicator,
+    ) -> Result<(), Self::Error>;
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ClockState {
+    StartupBlank,
+    // Needed when implementing frequency backups
+    #[allow(dead_code)]
+    StartupFreq,
+    MeasureFreq,
+    Spike,
+    Sync,
+}
+
+/// Controller responsible for actually
+/// deciding which adjustments to make based
+/// on results from the filtering and
+/// combining algorithms.
+#[derive(Debug, Copy, Clone)]
+pub struct ClockController<C: NtpClock> {
+    clock: C,
+    state: ClockState,
+    last_update_time: NtpInstant,
+    preferred_poll_interval: PollInterval,
+    poll_interval_counter: i32,
+    offset: NtpDuration,
+}
+
+pub enum ClockUpdateResult {
+    Ignore,
+    Step,
+    Slew,
+    Panic,
+}
+
+impl<C: NtpClock> ClockController<C> {
+    pub fn new(clock: C) -> Self {
+        clock.set_freq(0.).expect("Unable to set clock frequency");
+        Self {
+            clock,
+            state: ClockState::StartupBlank,
+            last_update_time: NtpInstant::now(),
+            preferred_poll_interval: PollInterval::MIN,
+            poll_interval_counter: 0,
+            offset: NtpDuration::ZERO,
+        }
+    }
+
+    // Preferred ratio between measured offset
+    // and measurement jitter
+    const POLL_FACTOR: i8 = 4;
+    // Threshold for changing desired poll interval
+    const POLL_ADJUST: i32 = 30;
+
+    pub fn update(
+        &mut self,
+        offset: NtpDuration,
+        jitter: NtpDuration,
+        root_delay: NtpDuration,
+        root_dispersion: NtpDuration,
+        leap_status: NtpLeapIndicator,
+        last_peer_update: NtpInstant,
+    ) -> ClockUpdateResult {
+        // Check that we have a somewhat reasonable result
+        if self.offset_too_large(offset) {
+            return ClockUpdateResult::Panic;
+        }
+
+        // Main decision making
+        //
+        // Combined, this code is responsible for:
+        //  - Filtering large but temporary spikes in the measured
+        //    offset to our timeservers
+        //  - Stepping the clock if a large difference persists long
+        //    enough
+        //  - Ensuring a proper initial frequency measurement on startup
+        //  - Making small (gradual) adjustments to the clock when we
+        //    only have a small error
+        if offset.abs() > NtpDuration::STEP_THRESHOLD {
+            // Large spikes are filtered initialy (to handle weird but temporary network issues)
+            // and then handled by stepping if they persist.
+            match self.state {
+                ClockState::Sync => {
```

Comment:
`use ClockState::*` would shorten these matches a bit

Reply:
I prefer being explicit with these and not polluting the namespace